python/pyiceberg/expressions/literals.py

```diff
@@ -58,25 +60,25 @@
     timestamp_to_micros,
     timestamptz_to_micros,
 )
-from pyiceberg.utils.singleton import Singleton
 
 T = TypeVar("T")
 
 
 class Literal(Generic[T], ABC):
     """Literal which has a value and can be converted between types"""
 
-    def __init__(self, value: T, value_type: type):
+    def __init__(self, value: T, value_type: Type):
         if value is None or not isinstance(value, value_type):
             raise TypeError(f"Invalid literal value: {value} (not a {value_type})")
         self._value = value
 
     @property
     def value(self) -> T:
-        return self._value  # type: ignore
+        return self._value
 
+    @singledispatchmethod
     @abstractmethod
-    def to(self, type_var) -> Literal:
+    def to(self, type_var: IcebergType) -> Literal:
```

Comment:
I added types to revive the TypeChecker, and it failed on this one. `to` returns a literal, therefore `AboveMax` and `BelowMin` also need to be literal. We can also turn this into a `Union[Literal, AboveMin, AboveMax]`. @rdblue thoughts?

Reply:
I think this PR improves the typing since we return a `Literal` in all the cases (or raise an exception).

To get this fully working, I think we need to add type information to the IcebergType, and then we can do the following:
```suggestion
    def to(self, type_var: IcebergType[T]) -> Literal[T]:
```